compiler/bcq-tools/generate_bcq_metadata

```diff
@@ -0,0 +1,214 @@
+
+            if infoname not in allowed_info_names:
+                is_valid = False
+                break
+
+            if prefix not in prefix_infonames_dict:
+                prefix_infonames_dict[prefix] = set()
+
+            prefix_infonames_dict[prefix].add(infoname)
+
+    # All the number of BCQ information should be same
+    num_of_bcqinfo = -1
+    for key in prefix_infonames_dict:
+        infonames = prefix_infonames_dict[key]
+        if num_of_bcqinfo == -1:
+            num_of_bcqinfo = len(infonames)
+        elif num_of_bcqinfo != len(infonames):
+            is_valid = False
+
+    # The number of BCQv1 information should be 6 or 7
+    if num_of_bcqinfo != 6 and num_of_bcqinfo != 7:
+        is_valid = False
+
+    # If BCQ information is invalid, return original model
+    if is_valid == False:
+        return original_graph_def
+
+    new_graph_def = tf.compat.v1.GraphDef()
+    for node in original_graph_def.node:
+        new_node = new_graph_def.node.add()
+        new_node.CopyFrom(node)
+
+    # Generate metadata header
+    metadata_values = generate_metadata_header(original_graph, 1, flags.output_arrays)
+
+    # Append metadata of BCQv1
+    metadata_values = np.append(metadata_values, num_of_bcqinfo + 1)
+
+    # Finish generating metadata
+    metadata_values = np.append(metadata_values, ONE_END_MAGICNUM)
+
+    # Generate metadata tensor
+    metadata_tensor = tf.make_tensor_proto(metadata_values, tf.int32)
+
+    new_node = new_graph_def.node.add()
+    new_node.op = "Const"
+    new_node.name = "one_compiler/bcqinfo_one_metadata"
+    new_node.attr["dtype"].CopyFrom(
+        tf.core.framework.attr_value_pb2.AttrValue(type=tf.int32.as_datatype_enum))
+    new_node.attr["value"].tensor.CopyFrom(metadata_tensor)
+    return new_graph_def
+
+
+def determine_bcq_version(flags):
+    """
+    CAUTION : For now, BCQ has only one version and thus always returns 1 when BCQ
+    information nodes are included. If new BCQ version is introduced,
+    this function must be updated accordingly.
+
+    When BCQ information does not exist, -1 is returned.
+    """
+
+    bcq_version = -1
+
+    original_graph = load_graph(flags.input_path)
+    original_graph_def = original_graph.as_graph_def()
+
+    for node in original_graph_def.node:
+        if node.op == "Const" and "/bcqinfo_" in node.name:
+            bcq_version = 1
+            break
+
+    return bcq_version
+
+
+def generate_bcq_metadata(flags):
+    """
+    Basic format of metadata is as following.
+        - Magic number indicating start
+        - Version of BCQ Format
+        - The number of original outputs
+        - Metadata based on each BCQ format
+        - Magic number indicating end
+    """
+    version = determine_bcq_version(flags)
+
+    # When there is not BCQ information, do nothing
+    if version == 1:
+        result_graph_def = generate_bcq_metadata_v1(flags)
+    else:
+        result_graph_def = load_graph(flags.input_path)
```

Comment:
I think it had better exit the script if `version != 1` rather than make the same model.

Reply:
Because of users of `tf2nnpkg`, we should make same model :(
More specifically, users both who want to apply BCQ and who do not know BCQ, they use same command as following.
```sh
tf2nnpkg --graphdef model.pb --info model.info -o .
```
Because of this situation, when a model has no BCQ information, we should copy original model.

Reply:
I think this user scenatio needs validation for `version != 1`

- In the future: when there is a BCQ file with version 2 or above and run with this version of script: what should we do?

Reply:
> I think this user scenatio needs validation for version != 1

If there is something I misunderstand, please let me know :)

At first, I wrote a code to throw an error when version or something related with BCQ is invalid.
However, I modified all the codes not to do anything after I received a review, "Just do nothing when BCQ information is invalid seems better than throwing an error. When model without BCQ is invalid, it should be died. However, when a model with BCQ is wrong, we can ignore only BCQ and generate a model without BCQ".
Following code is also written according to the policy.
https://github.com/Samsung/ONE/blob/c63a98f465d0c4c9854cbb37ade23bf22eddd2e3/compiler/luci/pass/src/FuseBCQPass.cpp#L508

So, should we change the policy?

---

> In the future: when there is a BCQ file with version 2 or above and run with this version of script: what should we do?

When new BCQ version is introduced, related implementations in `luci`, `bcq-tools` and so on should be done in advance of course.

Reply:
> Because of this situation, when a model has no BCQ information, we should copy original model.

I understand we should copy original model when a model has no BCQ information. But that does not mean that the tool must copy original model when the model has wrong version.

IMHO, a program should exit with error messages when its input is **invalid**. To me, it seems that a model without BCQ information is **valid**, while a model with wrong version is **invalid**.

It's my personal opinion (I'm not familiar with BCQ), so you can ignore it. :)

Reply:
> When new BCQ version is introduced, related implementations in luci, bcq-tools and so on should be done in advance of course.

Not that case. Please read again.

Reply:
Case 1: model version == script version: OK
Case 2: model version < script version: will be handled in the future version of script
Case 3: model version > script version: show message("cannot handle higher version") and stop maybe?

Reply:
@seanshpark 

> Case 2: model version < script version: will be handled in the future version of script

This is OK either. When script version is 3, it means that the script can handle BCQ version 1,2,3.

> Case 3: model version > script version: show message("cannot handle higher version") and stop maybe?

I have no objection for showing some messages and stopping :)
However as I wrote in https://github.com/Samsung/ONE/pull/4420#discussion_r495758729, I received an opinion, **Do not stop and just ignore BCQ**, and have written related codes according to the policy.
What I am curious is, **should I write a code to stop for such cases from now on** :-) ?

Reply:
> I received an opinion, Do not stop and just ignore BCQ

Maybe we can ask the person that gave you this opinion :)

Reply:
Hmm..... frankly speaking.... it was you O.O;;
Then I will follow to showing message and stop...!